File: web/lib/scanCoverage.ts
/** Coverage + snapshot picker for UW universe scans.

A completed NDX pass that 429'd every name still lands as
`candidates_found: 0`. Readers must not treat that as a real empty tape.
*/

export type ScanCoverage = {
  tickers?: number;
  ok?: number;
  no_setup?: number;
  rate_limited?: number;
  errors?: number;
  completed?: number;
};

const MIN_COMPLETED_RATIO = 0.5;

export function payloadHasCandidates(
  data: Record<string, unknown> | null | undefined,
): boolean {
  if (data == null) return false;
  const found = Number(data.candidates_found ?? 0);
  if (Number.isFinite(found) && found > 0) return true;
  return Array.isArray(data.results) && data.results.length > 0;
}

export function isCoverageFailedScan(
  data: Record<string, unknown> | null | undefined,
): boolean {
  if (payloadHasCandidates(data)) return false;
  if (data == null) return true;
  const coverage = data.coverage;
  let tickers = Number(data.tickers_scanned ?? 0);
  let completed = 0;
  if (coverage != null && typeof coverage === "object") {
    const block = coverage as ScanCoverage;
    tickers = Number(block.tickers ?? tickers);
    completed = Number(block.completed ?? 0);
  }
  if (!Number.isFinite(tickers) || tickers <= 0) return true;
  if (!Number.isFinite(completed) || completed < 0) return true;
  return completed / tickers < MIN_COMPLETED_RATIO;
}

export function pickUsableScanSnapshot(
  rows: Array<{ scan_time: string; payload: string }>,
): { scanTime: string; data: Record<string, unknown> } | null {
  const parsed: Array<{ scanTime: string; data: Record<string, unknown> }> = [];
  for (const row of rows) {
    try {
      const data = JSON.parse(row.payload) as Record<string, unknown>;
      if (data != null && typeof data === "object") {
        parsed.push({ scanTime: row.scan_time, data });
      }
    } catch {
      continue;
    }
  }
  return parsed.find((row) => payloadHasCandidates(row.data)) ?? parsed[0] ?? null;
}

File: scripts/utils/scan_coverage.py
"""Coverage + persist gate for UW universe scans.

A completed NDX pass that found nothing because UW 429'd every name still
writes ``candidates_found: 0``. Readers and writers must not treat that as
a real empty tape — it clobbers the last good snapshot.
"""

from __future__ import annotations

from typing import Any, Mapping, Optional

MIN_COMPLETED_RATIO = 0.5


def payload_has_candidates(payload: Optional[Mapping[str, Any]]) -> bool:
    if not isinstance(payload, Mapping):
        return False
    try:
        found = int(payload.get("candidates_found") or 0)
    except (TypeError, ValueError):
        found = 0
    if found > 0:
        return True
    results = payload.get("results")
    return isinstance(results, list) and len(results) > 0


def coverage_block(
    *,
    tickers: int,
    ok: int,
    no_setup: int,
    rate_limited: int,
    errors: int,
) -> dict[str, int]:
    return {
        "tickers": tickers,
        "ok": ok,
        "no_setup": no_setup,
        "rate_limited": rate_limited,
        "errors": errors,
        "completed": ok + no_setup,
    }


def is_coverage_failed(payload: Optional[Mapping[str, Any]]) -> bool:
    if payload_has_candidates(payload):
        return False
    if not isinstance(payload, Mapping):
        return True
    coverage = payload.get("coverage")
    try:
        tickers = int(payload.get("tickers_scanned") or 0)
        completed = 0
        if isinstance(coverage, Mapping):
            tickers = int(coverage.get("tickers") or tickers)
            completed = int(coverage.get("completed") or 0)
    except (TypeError, ValueError):
        return True
    if tickers <= 0:
        return True
    return completed / tickers < MIN_COMPLETED_RATIO


def should_persist_scan(
    payload: Optional[Mapping[str, Any]],
    prior: Optional[Mapping[str, Any]],
) -> bool:
    if payload_has_candidates(payload):
        return True
    if payload_has_candidates(prior):
        return False
    return not is_coverage_failed(payload)
